Fix OAuthBearer OIDC preprocessor directive when using CMake without CURL

emasab opened the pull request:
## Summary

Fixes several issues that surface when building librdkafka with CMake and a reduced
feature set, and adds CI coverage so these build-configuration regressions are caught
going forward.

Closes #5135, #5282

## Changes

- **Fix OAuthBearer OIDC preprocessor directive when building with CMake without CURL**
  (#5135) — corrects the conditional so compilation succeeds when CURL is disabled.
- **Add build-configuration CI checks** that build with CMake on Alpine and manylinux
  images with the optional features disabled, to catch compile/runtime breakage in
  minimal builds. Adds `packaging/tools/build-configurations-checks.sh`,
  `run-in-docker.sh`, `Dockerfile.alpine`, `Dockerfile.manylinux`, and a Semaphore job.
- **Skip test 0066 (plugins) when `dlopen` is unsupported**, so the broker-less quick
  suite passes in builds without dynamic loading.
- **Fix `rd_atomic{32,64}_set` returning the new value instead of the previous one in
  CMake builds.** CMake defined `HAVE_ATOMICS_{32,64}` (and optionally `_SYNC`) but never
  `HAVE_ATOMICS_{32,64}_ATOMIC`, so the setters fell through to a non-atomic fallback that
  returned the new value. This broke the planned-disconnection all-brokers-down handling,
  which relies on `rd_atomic32_set(&rkb_down_reported, 1)` returning the previous value to
  detect the `0 -> 1` transition — so `ALL_BROKERS_DOWN` was never raised and test 0095
  timed out under CMake. CMake now detects/defines `HAVE_ATOMICS_*_ATOMIC` (mirroring
  mklove's `configure.atomics`), and both setters select the `__atomic` / `__sync`
  implementation from the flags with the mutex path as a correct fallback.
- **Pin specific tool/image versions** for reproducible build-configuration checks.

## Testing

The new build-configuration checks run tests 0095 and 0121 under CMake/manylinux with all
optional features disabled; both pass (0095 previously timed out before the atomics fix).

emasab committed:
Fix compilation with CMake and without CURL

emasab committed:
Run build configuration tests with cmake,
with alpine or manylinux

emasab committed:
Skip tests 0066 if dlopen is unsupported

emasab committed:
Set specific version

emasab committed:
Fix rd_atomic{32,64}_set returning the new value in CMake builds

CMake builds defined HAVE_ATOMICS_{32,64} (and optionally _SYNC) but never
HAVE_ATOMICS_{32,64}_ATOMIC, so rd_atomic32_set()/rd_atomic64_set() fell
through to a non-atomic fallback that returned the new value instead of the
previous one.

This broke the planned-disconnection all-brokers-down handling, which relies
on rd_atomic32_set(&rkb_down_reported, 1) returning the previous value to
detect the 0 -> 1 transition. As a result the ALL_BROKERS_DOWN event was never
raised in CMake builds (test 0095 timed out).

- Detect and define HAVE_ATOMICS_{32,64}_ATOMIC in the CMake build when the
  __atomic builtins are available, mirroring mklove's configure.atomics.
- Rework rd_atomic{32,64}_set to select the __atomic / __sync implementation
  from the HAVE_ATOMICS_* flags and use the mutex path as fallback instead of
  a non-atomic store.

emasab committed:
Add changelog entries for CMake CURL and atomics fixes

Document the OAuthBearer OIDC/CURL CMake compilation fix (#5135) and the
rd_atomic{32,64}_set previous-value fix that restores ALL_BROKERS_DOWN under
CMake (#5282).

prashah-confluent commented on src/rdatomic.h:
```diff
@@ -119,19 +119,17 @@ static RD_INLINE int32_t RD_UNUSED rd_atomic32_set(rd_atomic32_t *ra,
-#elif !HAVE_ATOMICS_32
+#elif HAVE_ATOMICS_32 && HAVE_ATOMICS_32_ATOMIC
+        return __atomic_exchange_n(&ra->val, v, __ATOMIC_SEQ_CST);
+#elif HAVE_ATOMICS_32 && HAVE_ATOMICS_32_SYNC
```
Isn't this branch a dead code in case of CMake and eventually ending up using `#else` branch which is safe though because lock would be defined given HAVE_ATOMICS_32 is 0? 

CMake ends up setting only HAVE_ATOMICS_32_SYNC [reference](https://github.com/confluentinc/librdkafka/blob/c61cf7faee607fc46cbc9bdf0c883bd32b151e15/packaging/cmake/try_compile/rdkafka_setup.cmake#L59-L79) and not HAVE_ATOMICS_32. They both seem mutually exclusive in case of CMake. So this branch would never be entered right?

Should we change to something like below which seems to handle all scenarios and isn't based on the invariant `HAVE_ATOMICS_32_ATOMIC`

```
static RD_INLINE int32_t RD_UNUSED rd_atomic32_set(rd_atomic32_t *ra,
                                                   int32_t v) {
#ifdef _WIN32
        return InterlockedExchange((LONG *)&ra->val, v);
#elif HAVE_ATOMICS_32_SYNC
        return __sync_lock_test_and_set(&ra->val, v);
#elif !HAVE_ATOMICS_32
        int32_t r;
        mtx_lock(&ra->lock);
        r       = ra->val;
        ra->val = v;
        mtx_unlock(&ra->lock);
        return r;
#else
        return __atomic_exchange_n(&ra->val, v, __ATOMIC_SEQ_CST);
#endif
}
```

The only tradeoff I see is that default assumes `__atomic` is available.

emasab replied:
I was uniforming cmake in this PR to set the same values as mklove, so `HAVE_ATOMICS_32` for example is set to 1 in both cases and then either `HAVE_ATOMICS_32_ATOMIC` or `HAVE_ATOMICS_32_SYNC` is set to 1.
Just there's a case when using `try_compile(HAVE_ATOMICS_32_SYNC` where `HAVE_ATOMICS_32` isn't set to true. I've pushed the cmake change. In mklove instead there are cases where `HAVE_ATOMICS_32_SYNC` or `HAVE_ATOMICS_32_ATOMIC` aren't defined if `HAVE_ATOMICS_32` is false.

emasab committed:
Fix rd_atomic64_cas preprocessor branch ordering for available atomics

The #elif chain placed the mutex fallback (!HAVE_ATOMICS_64) before the
__atomic and __sync implementations, so the native atomic CAS branches were
never reachable and an incorrect non-atomic fallback was used. Reorder the
branches so HAVE_ATOMICS_64_ATOMIC and HAVE_ATOMICS_64_SYNC are selected
first, with the mutex-based path as the final fallback.

Also ignore the new share_consumer_commit_async and share_consumer_commit_sync
example binaries.

emasab committed:
set HAVE_ATOMICS_32 and HAVE_ATOMICS_64 to yes when one the options is available.

emasab committed:
Fix flaky do_test_fast_metadata_refresh by widening metadata request tolerance

Variation 1 (retries) of do_test_fast_metadata_refresh expects ~7 Metadata
requests but could occasionally observe more, causing intermittent failures
with the previous expected+1 (8) upper bound.

The count comes from two independent, timing-dependent sources:
- The fast-leader-query loop ("partition leader query"), driven by an
  exponential backoff timer with +/-20% jitter (rd_kafka_timer_exp_backoff),
  normally fires 6 times within the ~3s window but jitter can let a 7th fire
  land inside it.
- A "connected" metadata request issued when a broker connection transitions
  to UP (rd_kafka_broker_connect_up); with 3 mock brokers connecting at
  slightly different times a second one can land within the window.

These variances can stack, so the count can legitimately reach 9. Widen the
upper bound from expected+1 to expected+2 to account for it.

prashah-confluent reviewed (approved):
LGTM !! Thanks a lot @emasab
